src/relations/postgresql_provider.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Postgres client relation hooks & helpers."""
+
+
+import logging
+
+from charms.data_platform_libs.v0.database_provides import (
+    DatabaseProvides,
+    DatabaseRequestedEvent,
+)
+from charms.postgresql_k8s.v0.postgresql import (
+    PostgreSQLCreateDatabaseError,
+    PostgreSQLCreateUserError,
+    PostgreSQLDeleteUserError,
+    PostgreSQLGetPostgreSQLVersionError,
+)
+from ops.charm import CharmBase, RelationBrokenEvent
+from ops.framework import Object
+from ops.model import BlockedStatus
+
+from constants import DATABASE_PORT
+from utils import new_password
+
+logger = logging.getLogger(__name__)
+
+
+class PostgreSQLProvider(Object):
+    """Defines functionality for the 'provides' side of the 'postgresql-client' relation.
+
+    Hook events observed:
+        - database-requested
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str = "database") -> None:
+        """Constructor for PostgreSQLClientProvides object.
+
+        Args:
+            charm: the charm for which this relation is provided
+            relation_name: the name of the relation
+        """
+        self.relation_name = relation_name
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.charm = charm
+
+        # Charm events defined in the database provides charm library.
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_database_requested
+        )
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
+        """Generate password and handle user and database creation for the related application."""
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # Retrieve the database name and extra user roles using the charm library.
+        database = event.database
+        extra_user_roles = event.extra_user_roles
+
+        try:
+            # Creates the user and the database for this specific relation.
+            user = f"relation_id_{event.relation.id}"
```

Comment:
as per [this](https://chat.canonical.com/canonical/pl/1tp4gkzf1i8ft8z6c1gpnksfga) conversation, we should stick with:

```suggestion
            user = f"relation-{event.relation.id}"
```

Reply:
Fixed in https://github.com/canonical/postgresql-operator/pull/15/commits/d1a0911f6ce90104c34b55d2eaea4f4e68bcb3ea,